keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/Differentiator.java

```diff
@@ -1,165 +1,143 @@
-                }
+        Vertex<?> visiting;
+        while ((visiting = priorityQueue.poll()) != null) {
+
+            if (wrt.contains(visiting)) {
+                wrtOf.put(visiting.getId(), dwrtOf.get(visiting));
+                continue;
             }
-        }
 
-        return ofWrt;
-    }
+            if (!visiting.isProbabilistic()) {
 
-    public static <V extends Vertex & Differentiable> PartialDerivatives forwardModeAutoDiff(V vertex) {
-        Map<Vertex, PartialDerivatives> partialsOf = new HashMap<>();
-        Deque<V> stack = new ArrayDeque<>();
-        stack.push(vertex);
+                if (visiting instanceof Differentiable) {
 
-        while (!stack.isEmpty()) {
+                    Differentiable visitingDifferentiable = ((Differentiable) visiting);
+                    PartialDerivative derivativeOfOutputWrtVisiting = dwrtOf.get(visiting);
 
-            V head = stack.peek();
-            Set<Vertex> parentsThatPartialIsNotCalculated = parentsThatPartialIsNotCalculated(partialsOf, head.getParents());
+                    if (derivativeOfOutputWrtVisiting != null) {
+
+                        Map<Vertex, PartialDerivative> partialDerivatives = visitingDifferentiable.reverseModeAutoDifferentiation(derivativeOfOutputWrtVisiting);
+                        collectPartials(partialDerivatives, dwrtOf);
+
+                        for (Vertex parent : visiting.getParents()) {
+                            if (!alreadyQueued.contains(parent) && parent instanceof Differentiable) {
+                                priorityQueue.offer(parent);
+                                alreadyQueued.add(parent);
+                            }
+                        }
 
-            if (parentsThatPartialIsNotCalculated.isEmpty()) {
-                V top = stack.pop();
-                PartialDerivatives dTopWrtInputs = top.forwardModeAutoDifferentiation(partialsOf);
-                partialsOf.put(top, dTopWrtInputs);
-            } else {
-                for (Vertex parent : parentsThatPartialIsNotCalculated) {
-                    if (parent instanceof Differentiable) {
-                        stack.push((V) parent);
                     }
                 }
+
             }
         }
 
-        return partialsOf.get(vertex);
+        return new PartialsOf(ofVertex, wrtOf);
+    }
+
+    private static void ensureGraphValuesAndShapesAreSet(Vertex<?> vertex) {
+        vertex.getValue();
     }
 
-    private static Set<Vertex> parentsThatPartialIsNotCalculated(Map<Vertex, PartialDerivatives> partialsOf, Collection<? extends Vertex> parents) {
-        Set<Vertex> notCalculatedParents = new HashSet<>();
-        for (Vertex next : parents) {
-            if (!partialsOf.containsKey(next) && next instanceof Differentiable) {
-                notCalculatedParents.add(next);
+    private static void collectPartials(Map<Vertex, PartialDerivative> partialDerivatives,
```

Comment:
So the broadcast/reduce to scalar stuff is now handled in the vertices themselves (with help from the AutoDiffBroadcast class) now?  If so, much cleaner than handling that here IMO - good change.

Reply:
Yup. It's no longer the responsibility of the autodiff algo to check and handle these shape change issues. The operation logic is in a much better place to handle any shape change.

Reply:
👍